File: src/core/exiftool.js
import { execa } from "execa";

/**
 * Checks whether exiftool is available on the system PATH.
 *
 * @returns {Promise<boolean>}
 */
export async function isAvailable() {
  try {
    await execa("exiftool", ["-ver"]);
    return true;
  } catch {
    return false;
  }
}

/**
 * Reads metadata from files using exiftool.
 *
 * @param {string[]} filePaths - Paths to the files to read metadata from
 * @param {string[]} tags - Metadata tags to extract (e.g., ["EXIF:DateTimeOriginal"])
 * @returns {Promise<Map<string, object>>} Map of file paths to their metadata
 */
export async function readMetadata(filePaths, tags = []) {
  const args = [
    "-api",
    // Enable support for files >2GB
    "largefilesupport=1",
    // Output metadata as JSON
    "-json",
    // Print values as numbers/dates instead of human-readable format
    "-n",
    // Print group names with each tag (e.g., "EXIF:DateTimeOriginal")
    "-G",
    // Increase processing speed by not parsing all metadata
    "-fast",
    // Specify which metadata tags to extract
    ...tags.map((t) => `-${t}`),
    // Read file list from stdin
    "-@",
    // Read from stdin
    "-",
  ];

  try {
    const { stdout } = await execa("exiftool", args, {
      // Pipe filenames to stdin.
      input: filePaths.join("\n"),

      // 128MB buffer for safety.
      maxBuffer: 128 * 1024 * 1024,
    });

    const metadata = JSON.parse(stdout);
    const metadataMap = new Map();
    for (const item of metadata) {
      metadataMap.set(item.SourceFile, item);
    }

    return metadataMap;
  } catch (error) {
    throw new Error(`Failed to read metadata using exiftool ${error.message}`);
  }
}

/**
 * Writes metadata tags to a single file using exiftool.
 *
 * @param {string} filePath - Path to the file to write metadata to
 * @param {string[]} tags - Tags to write (e.g., ["DateTimeOriginal=2023:01:01 12:00:00"])
 * @returns {Promise<void>}
 */
export async function writeMetadata(filePath, tags) {
  if (tags.length === 0) {
    return;
  }

  const args = [
    "-api",
    // Enable support for files >2GB
    "largefilesupport=1",
    // Don't create backup files
    "-overwrite_original",
    // Tags to write
    ...tags.map((t) => `-${t}`),
    // File to write to
    filePath,
  ];

  try {
    await execa("exiftool", args);
  } catch (error) {
    throw new Error(`Failed to write metadata using exiftool: ${error.message}`);
  }
}

File: src/pixelkasten/exiftool.py
"""
ExifTool subprocess wrapper — read and write metadata.

Extracted from exif.py and extended with write support (ported from
packages/core/src/core/exiftool.js). Provides the low-level subprocess
interface; higher-level parsing lives in exif.py and handlers.py.
"""

import json
import subprocess
from pathlib import Path


def check_exiftool() -> None:
    """
    Verify that exiftool is installed and available on PATH.

    Raises RuntimeError with a clear installation message if not found.
    """
    try:
        result = subprocess.run(
            ["exiftool", "-ver"],
            capture_output=True,
            text=True,
            timeout=10,
        )
        if result.returncode != 0:
            raise RuntimeError(
                "exiftool is installed but returned an error. Check your installation."
            )
    except FileNotFoundError:
        raise RuntimeError(
            "exiftool is not installed. Install it with: brew install exiftool"
        )


def read_metadata(
    file_paths: list[Path],
    tags: list[str] | None = None,
) -> dict[str, dict]:
    """
    Read metadata from files using exiftool.

    Args:
        file_paths: Paths to the files to read.
        tags: Metadata tags to extract (e.g., ["EXIF:DateTimeOriginal"]).
              If None or empty, reads default tags.

    Returns:
        Dict mapping SourceFile path to raw exiftool metadata dict.
    """
    if not file_paths:
        return {}

    args = [
        "exiftool",
        "-api",
        "largefilesupport=1",
        "-json",
        "-n",
        "-G",
        "-fast",
    ]

    if tags:
        for tag in tags:
            args.append(f"-{tag}")

    args.extend(["-@", "-"])

    stdin_text = "\n".join(str(p) for p in file_paths)

    result = subprocess.run(
        args,
        input=stdin_text,
        capture_output=True,
        text=True,
        timeout=300,
    )

    # exiftool returns 1 for minor warnings (missing tags), not errors.
    if result.returncode not in (0, 1):
        raise RuntimeError(f"Failed to read metadata using exiftool: {result.stderr}")

    if not result.stdout.strip():
        return {}

    entries = json.loads(result.stdout)
    return {entry.get("SourceFile", ""): entry for entry in entries}


def write_metadata(file_path: str | Path, tags: list[str]) -> None:
    """
    Write metadata tags to a file using exiftool.

    Args:
        file_path: Path to the file to write metadata to.
        tags: Tags to write (e.g., ["SubSecDateTimeOriginal=2023:05:20 14:30:00+00:00"]).
              If empty, returns immediately without calling exiftool.
    """
    if not tags:
        return

    args = [
        "exiftool",
        "-api",
        "largefilesupport=1",
        "-overwrite_original",
    ]

    for tag in tags:
        args.append(f"-{tag}")

    args.append(str(file_path))

    result = subprocess.run(
        args,
        capture_output=True,
        text=True,
        timeout=60,
    )

    if result.returncode != 0:
        raise RuntimeError(f"Failed to write metadata using exiftool: {result.stderr}")
